docs/architecture/adr-022-system-merge.md

```diff
@@ -0,0 +1,63 @@
+
+Having duplication requires both a broader knowledge base and more maintenance load for engineers. There is a larger surface area for bugs and gains from optimizations are not captured as efficiently as they could.
+
+Outside of the aforementioned tech debt, there is untapped potential in taking the benefits of LibP2P and porting it across to the Tendermint consensus algorithm. LibP2P has a more sophisticated and efficient gossiping protocol than Tendermint's. It also has a more robust peer discovery and peer management system.
+
+## Decision
+
+The decision is to converge to a LibP2P based network and to shape out reusable components that can be used by all node types: consensus, data and light. This is to be gradually done across four phases, each phase likely requiring it's own in-depth ADR:
+
+### Phase 1: Removal of the bridge node
+
+The first step involves porting `Init` and `Start` methods from `celestia-app` and allowing consensus nodes to be initialized and run through the `celestia` binary. This will allow for a argument to be passed in `Start` that enables the consensus nodes to directly push the extended data square (EDS) that is finalized in `FinalizeBlock` to the LibP2P based networks. Thus consensus nodes will temporarily be a hybrid of both networking systems and replace the need for a separate bridge node. Having a single binary will greatly simplify both the release process and end-to-end testing.
+
+### Phase 2: Migration of storage, syncing and fraud
+
+The next step migrates first the storage engine and then the syncing and fraud protocols (`go-header` and `go-fraud`) to `celestia-app`. This replaces the existing `BlockStore` and blocksync and evidence reactors. Streamlining the storage will require careful migration but will eliminate the duplication of storage and unify many of the data exchange protocols and the public APIs. Here, the consensus node can become more integrated with the `nodebuilder` package as it begins to be split up into a series of modules. This phase may also try to align configuration files and the general file directory.
+
+### Phase 3: Rewiring consensus, txpool and statesync
+
+The third phase consists of replacing the interface between the consensus, mempool and statesync protocols to Tendermint's p2p layer with LibP2P. This can be done in tandem with phase 2 and may be coupled with other refactors and performance improvements to those components. This will manifest itself as four new modules to `celestia-node`: consensus, mempool, statesync, and execution (a standalone Celestia state machine. This is somewhat similar to `app.App`).
+
+### Phase 4: Unification of Public APIs
+
+Lastly, legacy APIs like the RPC layer will be deprecated in favour of a unified public API. Every node should offer the same collection of endpoints and be capable of verifying all retrieved data.
+
+## Considerations
```

Comment:
Curious to know, does every phase mean a dedicated ADR with more details?

Reply:
Yes, this is a high level ADR. Each of these phases will require a more dedicated description of the design and their tradeoffs